### PR TITLE
And default and record-dot syntaxes

### PR DESCRIPTION
**Pull request by wingedtrashcan**
This PR implements nicer surface syntax for Nightfall in a way that allows the Haskell type checker to refute a variety a lot of ill-constructed Nightfall scripts. This is more of a proof-of-concept than a final implementation when it comes to safety checks and documentation, but it should be enough to judge whether we want such complexity in the surface language or not. In my opinion it is worth it, but YMMV.

Don't merge this PR. If we decide it's worth it, we need to turn existing tests into golden ones to make sure this PR actually doesn't change the programs.

### CODE REVIEW

File: src/Examples/Cond.hs [hunk 1]
```diff
@@ -51,15 +55,15 @@ ifVarBody = do
...
-        (ret $ varF "nokVal")
+    Felt <- declare.a 145
+    Felt <- declare.b 79
+    Felt <- declare.target 203
+    Felt <- declare.sum $ get.a + get.b
+    Felt <- declare.okVal 10
+    Felt <- declare.nokVal 20
+    ifElse (get.sum `eq` get.target)
+        (ret get.okVal)
+        (ret get.nokVal)
```
Comment (wingedtrashcan) [inline src/Examples/Cond.hs:66]: Actually, I'm not sure if I like the `get` name... I feel like `use` would be better:

```haskell
    <...>
    ifElse (use.sum `eq` use.target)
        (ret use.okVal)
        (ret use.nokVal)
```

Not sure.

Comment (wingedtrashcan) [reply src/Examples/Cond.hs:66]: Nevermind, we can't use `use`, because it has a meaning in Miden (~same as `import` in Haskell).

**Comment by wingedtrashcan** [comment]
I've added the default syntax which is some simple variation of HOAS. It turned out that HOAS has an advantage over the record-dot craze, namely that type inference is better. While I [can](https://github.com/qredo/nightfall/commit/882b2ca5d722cb648c1d52c043ccebb052f7c536) (I'll make a PR for that as a follow-up) make the craze infer types, it only works top-down (i.e. latter lines cannot influence type inference in former ones), because unlike with HOAS variables are connected very poorly type-wise, while in HOAS we have a literal term-level variable that drives type inference being used. It's a trade-off: with record-dot you don't need to create a term-level variable and so there is no shadowing of existing Haskell names for example (which is why I had to change the `length` Nightfall variable to `len` for example). Plus name duplication for HOAS is unfortunate too (one name on the Nightfall side and its counterpart on the Haskell side).

I'd definitely prefer record-dot over HOAS if type inference worked both ways with the former too, since otherwise record-dot is more expressive and ergonomic.

@allezlom13 ready for review.

**Comment by allezlom13** [comment]
One thought about the "duplicate names" (Haskell name + Nightfall name): how about we just kill the Nightfall one altogether? Is it essential? Could use unique integers and call it a day?

**Comment by wingedtrashcan** [comment]
> One thought about the "duplicate names" (Haskell name + Nightfall name): how about we just kill the Nightfall one altogether? Is it essential? Could use unique integers and call it a day?

Interesting idea... But then if we ever need to debug or even just look into a pretty-printed Nightfall program, it will be orders of magnitude harder to figure out what the program does. E.g. Nightfall programs pretty-printed into golden files will be unreadable and we won't be able to check correctness of diffs in a lot of cases, because if you just shuffle several orthogonal bindings around, you're already making it very hard to understand the rest of the program and introduce a lot of noise into the diff too (a variable name stays a variable name, but a unique number changes across all use sites if you shuffle binders).

So I guess I'd rather have duplicate names than risk running into this situation.

**Review by allezlom13 — APPROVED** [review]
LGTM

**Pull request merged**